src/maxreg.rs

```diff
@@ -0,0 +1,135 @@
+use crate::traits::{CmRDT, CvRDT};
+use serde::{Deserialize, Serialize};
+use std::{error, fmt};
+
+/// `MaxReg` Holds a monotonically increasing value that implements the Ord trait. For use of floating-point values,
+/// you must create a wrapper (or use a crate like `float-ord`)
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub struct MaxReg<V: Ord + Copy> {
+    /// `val` is the opaque element contained within this CRDT
+    /// Because `val` is monotonic, it also serves as a marker and preserves causality
+    pub val: V,
+}
+
+impl<V: Default + Ord + Copy> Default for MaxReg<V> {
+    fn default() -> Self {
+        Self { val: V::default() }
+    }
+}
+
+/// The Type of validation errors that may occur for an MaxReg.
+#[derive(Debug, PartialEq)]
+pub enum Validation {
+    /// A conflicting change to a MaxReg CRDT is when the value being validated is less-than self.val
+    ConflictingValue,
+}
+
+impl error::Error for Validation {
+    fn description(&self) -> &str {
+        match self {
+            Validation::ConflictingValue => {
+                "Comparison failed! `val` should be monotonically increasing."
+            }
+        }
+    }
+}
+
+impl fmt::Display for Validation {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+impl<V: Ord + Copy> CvRDT for MaxReg<V> {
+    type Validation = Validation;
+    /// Validates whether a merge is safe to perfom
+    ///
+    /// Returns an error if `val` is greater than the proposed new value
+    /// ```
+    /// use crdts::{maxreg, MaxReg, CvRDT};
+    /// let mut ub_1 = MaxReg { val: 1};
+    /// let ub_2 = MaxReg { val: 0 };
+    /// // errors!
+    /// assert_eq!(ub_1.validate_merge(&ub_2), Err(maxreg::Validation::ConflictingVal));
+    /// ```
+    fn validate_merge(&self, other: &Self) -> Result<(), Self::Validation> {
+        self.validate_update(&other.val)
+    }
+
+    /// Combines two `MaxReg` instances according to the value that is greatest
+    fn merge(&mut self, MaxReg { val }: Self) {
+        self.update(val)
+    }
+}
+
+impl<V: Ord + Copy> CmRDT for MaxReg<V> {
+    // MaxRegs's are small enough that we can replicate
+    // the entire state as an Op
+    type Op = Self;
+    type Validation = Validation;
+
+    fn validate_op(&self, op: &Self::Op) -> Result<(), Validation> {
+        self.validate_update(&op.val)
+    }
```

Comment:
We don't need validation for this CRDT, there's no such thing as an invalid operation in for this CRDT.

just return `Ok(())`

Reply:
Can change this to be more simple. I know it can never happen, but thought it could also be used as a convenient test or debugging method; if you are wanting to test some value you know should be larger than the current value, but you validate it isn't, then you might want to simply call `validate_update` on the incoming value you are testing.

Reply:
You wouldn't reuse the `validate_update()` API for that. The validate API is used for the "causal" class of CRDT's that depend on "in-order" delivery of operations. We don't want to dilute the meaning of this validation API by adding more use-cases to it.